File: tests/bench/spectral/spectral.py
import os

_LEVEL = os.environ.get("BENCH_LEVEL", "hard")
SCALE = 16 if _LEVEL == "easy" else 4 if _LEVEL == "medium" else 1
# The matrix size stays; the repetitions around it are what the level shortens.
REPS = max(1, 6 // SCALE)


def evalA(i, j):
    s = i + j
    return 1.0 / float(s * (s + 1) // 2 + i + 1)


def timesA(v, n):
    out = []
    for i in range(n):
        total = 0.0
        for j in range(n):
            total += evalA(i, j) * v[j]
        out.append(total)
    return out


def timesAt(v, n):
    out = []
    for i in range(n):
        total = 0.0
        for j in range(n):
            total += evalA(j, i) * v[j]
        out.append(total)
    return out


n = 550
u = [1.0] * n
v = []
for _r in range(REPS):
    v = timesAt(timesA(u, n), n)
    u = timesAt(timesA(v, n), n)

vBv = 0.0
vv = 0.0
for i in range(n):
    vBv += u[i] * v[i]
    vv += v[i] * v[i]
print(n)
print(f"{(vBv / vv) ** 0.5:.9f}")

File: tests/bench/spectral/spectral.cpp
// Same program as spectral.jai. See tests/bench/README.md for why the C++ and
// Java rows are here and what they are not.
#include <cmath>
#include <cstdint>
#include <cstdio>
#include <vector>

#include <cstdlib>
#include <cstring>

// BENCH_LEVEL picks how much work this benchmark does: easy a sixteenth of it,
// medium a quarter, hard (the default, and anything unrecognised) all of it.
static long bench_scale() {
    const char *l = std::getenv("BENCH_LEVEL");
    if (l && std::strcmp(l, "easy") == 0) return 16;
    if (l && std::strcmp(l, "medium") == 0) return 4;
    return 1;
}
static const long SCALE = bench_scale();
// A repetition count around a fixed matrix: never below one round.
static const long ROUNDS = (6 / SCALE) < 1 ? 1 : (6 / SCALE);

static double evalA(int64_t i, int64_t j) {
    int64_t s = i + j;
    return 1.0 / (double)(s * (s + 1) / 2 + i + 1);
}

static std::vector<double> timesA(const std::vector<double> &v, int n) {
    std::vector<double> out;
    for (int i = 0; i < n; i++) {
        double sum = 0.0;
        for (int j = 0; j < n; j++) sum += evalA(i, j) * v[(size_t)j];
        out.push_back(sum);
    }
    return out;
}

static std::vector<double> timesAt(const std::vector<double> &v, int n) {
    std::vector<double> out;
    for (int i = 0; i < n; i++) {
        double sum = 0.0;
        for (int j = 0; j < n; j++) sum += evalA(j, i) * v[(size_t)j];
        out.push_back(sum);
    }
    return out;
}

int main() {
    const int n = 550;
    std::vector<double> u((size_t)n, 1.0);
    std::vector<double> v;
    for (long r = 0; r < ROUNDS; r++) {
        v = timesAt(timesA(u, n), n);
        u = timesAt(timesA(v, n), n);
    }

    double vBv = 0.0;
    double vv = 0.0;
    for (int i = 0; i < n; i++) {
        vBv += u[(size_t)i] * v[(size_t)i];
        vv += v[(size_t)i] * v[(size_t)i];
    }
    std::printf("%d\n", n);
    std::printf("%.9f\n", std::pow(vBv / vv, 0.5));
    return 0;
}
